engine/src/org/pentaho/di/trans/steps/joinrows/JoinRows.java

```diff
@@ -446,7 +450,9 @@ public void dispose( StepMetaInterface smi, StepDataInterface sdi ) {
     // Remove the temporary files...
     if ( data.file != null ) {
       for ( int i = 1; i < data.file.length; i++ ) {
-        data.file[i].delete();
+        if ( data.file[i] != null ) {
+          data.file[i].delete();
+        }
```

Comment:
I have faced the issue when you don't define a temp directory or the path isn't found then the transformation hangs for ever. It is caused by NPE in finally block https://github.com/pentaho/pentaho-kettle/blob/master/engine/src/org/pentaho/di/trans/step/RunThread.java#L96